cunumeric/module.py

```diff
@@ -6227,3 +6227,719 @@ def bincount(
+        range(min_dim_index + num_axes, ndim)
+    )
+    reshaped = tuple(
+        [
+            collapsed_shape if k == min_dim_index else arr_shuffled.shape[k]
+            for k in redimed
+        ]
+    )
+
+    arr_reshaped = arr_shuffled.reshape(reshaped)
+    return (min_dim_index, arr_reshaped)
+
+
+# args:
+#
+# arr:      [in] source nd-array on which quantiles are calculated;
+#                preccondition: assumed sorted!
+# q_arr:    [in] quantile input values nd-array;
+# axis:     [in] axis along which quantiles are calculated;
+# method:   [in] func(q, n) returning (gamma, j),
+#                where = array1D.size;
+# keepdims: [in] boolean flag specifying whether collapsed axis
+#                should be kept as dim=1;
+# to_dtype: [in] dtype to convert the result to;
+# qs_all:   [in/out] result pass through or created (returned)
+# input_is_scalar: [in] specifies if q input was scalar
+# return: nd-array of quantile output values
+#         where its shape is obtained as:
+#         concatenating q_arr.shape with arr.shape \ {axis}
+#         (the shape of `arr` obtained by taking the `axis` out)
+#
+def quantile_impl(
+    arr,
+    q_arr,
+    axis,
+    axes_set,
+    original_shape,
+    method,
+    keepdims,
+    to_dtype,
+    qs_all,
+    input_is_scalar,
+):
+
+    ndims = len(arr.shape)
+
+    if axis is None:
+        n = arr.size
+        remaining_shape = []  # only `q_arr` dictates shape;
+        # quantile applied to `arr` seen as 1D;
+    else:
+        n = arr.shape[axis]
+
+        # arr.shape -{axis}; if keepdims use 1 for arr.shape[axis]:
+        # (can be empty [])
+        #
+        if keepdims:
+            remaining_shape = [
+                1 if k in axes_set else original_shape[k]
+                for k in range(0, len(original_shape))
+            ]
+        else:
+            remaining_shape = [
+                arr.shape[k] for k in range(0, ndims) if k != axis
+            ]
+
+    # quantile interpolation method to act on vectors,
+    # like an axpy call:
+    # [gamma](vleft, vright) { return (1-gamma)*vleft + gamma*vright;}
+    #
+    linear_interp = (
+        lambda gamma, arr_lvals, arr_rvals: (1.0 - gamma) * arr_lvals
+        + gamma * arr_rvals
+    )
+
+    # compose qarr.shape with arr.shape:
+    #
+    # result.shape = (q_arr.shape, arr.shape -{axis}):
+    #
+    qresult_shape = (*q_arr.shape, *remaining_shape)
+
+    # construct result NdArray, non-flattening approach:
+    #
+    res = zeros(qresult_shape, dtype=to_dtype)
+
+    for index, q in np.ndenumerate(q_arr):
+        (gamma, j) = method(q, n)
+        (left_pos, right_pos) = (j, j + 1)
+
+        # (N-1) dimensional ndarray of left, right
+        # neighbor values:
+        #
+        # non-flattening approach:
+        #
+        # singleton list: [left_pos];
+        # extract values at index=left_pos;
+        arr_1D_lvals = arr.take([left_pos], axis)
```

Comment:
You could also avoid the `reshape()` if you do take `arr.take(left_pos, axis)` here, and correspondingly for `arr_1D_rvals` below.

Reply:
I can replace the singleton `[pos]` with just `pos`, but it still fails w/o `reshape()`. Again, I get fake additional 1D errors, like this one:
```
ValueError: could not broadcast input array from shape (2,4) into shape (2,1,4)
```

Reply:
I know. That's bc/ `keepdims = True` introduces a fictitious dimension of 1 along selected axis (per `numpy` specs).

But `take(index, axis)` is not aware of that. And I don't know of a way to make it aware.

Reply:
This was addressed via https://github.com/nv-legate/cunumeric/commit/ea611a2269cb359a2965d177e0754bb1e07b3b79.